onnxruntime/core/session/inference_session.cc

```diff
@@ -159,6 +166,16 @@ Status VerifyEachNodeIsAssignedToAnEp(const Graph& graph, const logging::Logger&
       }
     }
   }
+
+  // If one preferred provider was requested, and others were used instead,
+  // warn the user of the the fallback which is otherwise a silent perf concern.
+  if (!node_placement_set.empty() && providers.NumProviders() == 1) {
+    if (node_placement_set.size() > 1 ||
```

Comment:
>(node_placement_set.size() > 1 [](http://example.com/codeflow?start=7&length=30)

Why is this check required?

Reply:
Dmitri thanks for looking - I wouldn't invest *too* much time into reviewing this yet, as it's just a draft reference for the sake of an email discussion with Faith/Pranav about how to handle these perf issues, but this is one *possible* route. I've adjusted the logic a bit.

In summary, if you ask for one provider, and you get a different one, or if you ask for just one and you get 2 back, then that's not expected. Unfortunately *currently* ORT always appends the CPU provider into the list, and so it's not possible to tell when the user meant ["DmlExecutionProvider", "CpuExecutionProvider"] (and yes, it's really okay to use both) or they meant ["DmlExecutionProvider"] (and they did not intend to get any CPU fallback at all, then being confused by the perf).

Reply:
(And I now have some pending changes to make after Pranav's reply on how we can improve the messaging...)

Reply:
It's updated now after feedback from Pranav, regarding explicit vs implicit EP registration (we don't want to bother them if they added the CPU provider anyway, only for implicit fallback, and when it happens, to advise them to rerun with verbose to verify expectations).

Reply:
@skottmckay Partitioning tools would be even more informative, but I feel the messaging to just rerun with verbose (Pranav's suggestion) suffices because it shows all the node assignments within `VerifyEachNodeIsAssignedToAnEp`. Even without partitioning logic, just knowing the operator is usually enough.